controllers/kubeadmconfig_controller.go

```diff
@@ -166,13 +131,83 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
 	}
 
+	// Store Config's state, pre-modifications, to allow patching
+	patchConfig := client.MergeFrom(config.DeepCopy())
+
+	// Check for control plane ready. If it's not ready then we will requeue the machine until it is.
+	// The infrastructure provider *must* set this value to use this bootstrap provider.
+	if cluster.Annotations[ControlPlaneReadyAnnotationKey] != "true" {
+		// if it's NOT a control plane machine, requeue
+		if !util.IsControlPlaneMachine(machine) {
+			log.Info("Control plane is not ready, requeing until ready.")
+			return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+		}
+
+		//TODO(fp) use init lock so only the first machine configured as control plane get processed, everything else gets requeued
+
+		//TODO(fp) check supported combinations of the InitConfiguration and ClusterConfiguration (only one, both, any?)
+
+		// if the machine has cluster and or init defined on it then generate the init regular join
+		if config.Spec.InitConfiguration != nil && config.Spec.ClusterConfiguration != nil {
+			log.Info("Creating BootstrapData for the init control plane")
+
+			// get both of these to strings to pass to the cloud init control plane generator
+			initdata, err := json.Marshal(config.Spec.InitConfiguration)
+			if err != nil {
+				log.Error(err, "failed to marshal init configuration")
+				return ctrl.Result{}, err
+			}
+			clusterdata, err := json.Marshal(config.Spec.ClusterConfiguration)
+			if err != nil {
+				log.Error(err, "failed to marshal cluster configuration")
+				return ctrl.Result{}, err
+			}
+
+			//TODO(fp) check what is the expected flow for certificates
+			certificates, _ := NewCertificates()
+
+			cloudInitData, err := cloudinit.NewInitControlPlane(&cloudinit.ControlPlaneInput{
+				//TODO(fp) if defined, cluster.Status.APIEndpoints[0] should override ClusterConfiguration.ControlPlaneEndpoint
+				InitConfiguration:    initdata,
+				ClusterConfiguration: clusterdata,
+				Certificates: cloudinit.Certificates{
+					CACert:           string(certificates.ClusterCA.Cert),
+					CAKey:            string(certificates.ClusterCA.Key),
+					EtcdCACert:       string(certificates.EtcdCA.Cert),
+					EtcdCAKey:        string(certificates.EtcdCA.Key),
+					FrontProxyCACert: string(certificates.FrontProxyCA.Cert),
+					FrontProxyCAKey:  string(certificates.FrontProxyCA.Key),
+					SaCert:           string(certificates.ServiceAccount.Cert),
+					SaKey:            string(certificates.ServiceAccount.Key),
+				},
+			})
+			if err != nil {
+				log.Error(err, "failed to generate cloud init for bootstrap control plane")
+				return ctrl.Result{}, err
+			}
+
+			config.Status.BootstrapData = cloudInitData
+			config.Status.Ready = true
+
+			if err := r.Update(ctx, &config); err != nil {
```

Comment:
Update or Patch?

Reply:
Patch please

Reply:
I think there was a rebase gone awry somewhere. This should use patch.

Reply:
@ncdc @chuckha There are some weird problem after this change.
It seems that every operation with `k8sClient` implemented in tthe test loses GVK, and this leads `r.Status().Patch` to fail with "the server could not find the requested resource (patch kubeadmconfigs.bootstrap.cluster.x-k8s.io control-plane-config)". 
Any suggestion?

Reply:
Known issue, 1 sec

Reply:
Copy what's in here: https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/pull/69/files

Reply:
@ncdc the problem is a little bit different
when I call `k8sClient.Create(context.Background(), config)` in tests, it seems the GVK gets stripped from config before being stored in etcd, or at least this is what I assume when I check config after calling of `k8sClient.Get(ctx, controlplaneConfigKey, config)`

Reply:
When apimachinery decodes into a typed struct, the decoder strips the TypeMeta from the object (the theory at the time being that because it was a typed object, you knew its API version, group, and kind). There's a WIP PR to k/k to fix this.

This sounds like the same issue - right?

Reply:
ok, for the time being, I commented on the behavioral tests that was failing;
considering that everything is covered by unit tests, IMO this can be addressed in follow up PRs when the root of this issue is fixed in k/k